src/unix/udp.c

```diff
@@ -616,23 +639,19 @@ static int uv__setsockopt_maybe_char(uv_udp_t* handle,
   if (val < 0 || val > 255)
     return -EINVAL;
 
-  if (handle->flags & UV_HANDLE_IPV6)
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IPV6,
-                   option6,
-                   &arg,
-                   sizeof(arg));
-  else
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IP,
-                   option4,
-                   &arg,
-                   sizeof(arg));
-
-  if (r)
-    return -errno;
+/*
+ * On Solaris and derivatives such as SmartOS, the length of socket options
+ * is always sizeof(int), except for the IP_MULTICAST_LOOP and IP_MULTICAST_TTL
+ * options.
+ */
+#if defined(__sun)
```

Comment:
I agree that the original change was confusing to read. However reading:
```
if ((handle->flags & UV_HANDLE_IPV6)
    uv__setsockopt(handle, IP_MULTICAST_TTL, IPV6_MULTICAST_HOPS, ttl, sizeof(ttl));
```
also felt a bit confusing, because even if we know that the handle identifies an IPv6 socket, we set both IPv4 and IPv6 options.

So I tried to come up with another refactoring of my change that takes another approach in making it easier to read. It's the second commit of this PR. Please let me know if that's a step in the right direction, or if you still think that your suggestion is better.

Thank you!